container-entrypoint.sh

```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
+# SPDX-License-Identifier: MIT
+#
+# Copyright (C) 2023 Olliver Schinagl <oliver@schinagl.nl>
+#
+# A beginning user should be able to docker run image bash (or sh) without
+# needing to learn about --entrypoint
+# https://github.com/docker-library/official-images#consistency
+
+set -eu
+
+bin='docker-gen'
+
+# run command if it is not starting with a "-" and is an executable in PATH
+if [ "${#}" -le 0 ] || \
+   [ "${1#-}" != "${1}" ] || \
+   [ -d "${1}" ] || \
+   ! command -v "${1}" > '/dev/null' 2>&1; then
+	entrypoint='true'
+fi
+
+exec ${entrypoint:+${bin:?}} "${@}"
```

Comment:
What's the rationale behind the `[ -d "${1}" ]` check (this would end up passing a folder as argument to docker-gen, and AFAIK this does not work) ?  🤔

While I understand the `command -v "${1}" > '/dev/null' 2>&1` check, I don't think that's the way to go because it would mean commands with typo would be passed as argument to docker-gen instead of failing with a `foo: not found` as they should. A simple `[ -f "${1}" ]` work in this case: executables in `PATH` do not check as files.

Finally the `exec ${entrypoint:+${bin:?}} "${@}"` is hard to decipher and needlessly complex, the way it's done in the official docker example is clearer.

Suggestion taking all of the above into account:

```shell
#!/bin/sh

set -eu

# if command starts with an "-" or is a file, pass it as argument(s) to docker-gen, else run it.
if [ "${#}" -le 0 ] || [ "${1#-}" != "${1}" ] || [ -f "${1}" ]; then
	set -- docker-gen "${@}"
fi

exec "${@}"
```

The `exit 0` at the end as no real use since we're using `set -e`.

Reply:
It looks like github ate my comment, as I'm very sure I replied in detail. I'll try to find it again.

Reply:
Aw crap. 😞

Reply:
Yep, gone; was hoping I could get it in the edit field on my laptop, no such luck ... so let me try again.

> What's the rationale behind the `[ -d "${1}" ]` check (this would end up passing a folder as argument to docker-gen, and AFAIK this does not work) ? thinking

I have to dig deep now, as I use this template for everything (to keep them all the same, nice and consistent). Afaik, if you have a container, that accepts a directory as input, lets say the `ls` container, if the directory exists, it'll allow you to pass the directory as argument to the binary. But you got me, I'm trying remember really hard what the usecase was (there was one :p)

> 
> While I understand the `command -v "${1}" > '/dev/null' 2>&1` check, I don't think that's the way to go because it would mean commands with typo would be passed as argument to docker-gen instead of failing with a `foo: not found` as they should. A simple `[ -f "${1}" ]` work in this case: executables in `PATH` do not check as files.

But was it a command you where trying to execute? or was it something you wanted to pass a long. If you assume the script to be dedicated only to `docker-gen` yeah, sure. But in a more generic case. Also, if I supply `cp` instead of `/usr/bin/cp` it should still work.

> 
> Finally the `exec ${entrypoint:+${bin:?}} "${@}"` is hard to decipher and needlessly complex, the way it's done in the official docker example is clearer.

Eye of the beholder, sure enough. I used to have `${bin:+${bin}}` before, is that easier on the eyes? I took this new approach, as I wanted to keep the bin definition at the top of the file, for easy recognition.

The docker approach is equally 'confusing'. What's with the `set --`, that is also less clear (doesnt' that eat the existing `set -eu` also? Idk, never had to use it yet). The reason it is a variable though; is to ensure a single exit point of the script.

> 
> Suggestion taking all of the above into account:
> 
> ```shell
> #!/bin/sh
> 
> set -eu
> 
> # if command starts with an "-" or is a file, pass it as argument(s) to docker-gen, else run it.
> if [ "${#}" -le 0 ] || [ "${1#-}" != "${1}" ] || [ -f "${1}" ]; then
>     set -- docker-gen "${@}"
> fi
> 
> exec "${@}"
> ```
> 
> The `exit 0` at the end as no real use since we're using `set -e`.
ah, but what if some hacker dude does `echo 'rm -rf /' >> script.sh`? It's a shell scripting best-practise (afaik its in the google bash styleguide even). I like best-practises :) 'always have a defined exit point at the end of your script', in case something goes haywire.